### PR TITLE
Replace version-based conditional code loading with dynamic dispatch

### PR DESCRIPTION
**Pull request by mofeing**
Superseeds #70.

From https://github.com/JuliaLabs/MLIR.jl/pull/70#issue-2259534168
> ... several projects experimenting with MLIR.jl, like [Coil.jl](https://github.com/pangoraw/Coil.jl) and [Reactant.jl](https://github.com/EnzymeAD/Reactant.jl), need to import the whole MLIR.jl code just because they use another libMLIR (included in XLA or IREE). But they just need a way to overwrite MLIR_jll.mlir_c.
>
> Since we conditionally load code based on Julia's LLVM version (mainly for MLIR.API and MLIR.Dialects), the loaded code won't match the version of the custom MLIR loaded version.

Also, different libraries in the same project might require different libMLIR.

The only solution I foresee is to:
1. Have all MLIR API versions available.
2. Dynamically dispatch MLIR functions based on a dynamically scoped `MLIR_VERSION` variable.

# Changes

- FFI calls to `libMLIR` are now performed on `MLIR_C_PATH` instead of `MLIR_jll.mlir_c`.
  - `MLIR_C_PATH` defaults to `MLIR_jll.mlir_c`, but can be "statically" changed using `Preferences`.
  - `MLIR_C_PATH` is a `ScopedValue`, so its value can be dinamically changed using `ScopedValues.with`.
- Removed static version-based conditional code loading, and move versioned-functions to "versioned modules" (e.g. `API.v14`).
  - This affects the definitions in `API` and `Dialects`.
- Type definitions transpiled from MLIR C headers have been refactored out of `API` versioned modules and now are shared between all versions.
- ~Calls to `API`/`Dialects` functions are version-based dynamically dispatched to the correct versioned module through a `VersionDispatcher` (`API.Dispatcher`, `Dialects.Dispatcher`).~
- Calls to MLIR-functions in `API` are dynamically dispatched to versioned modules using `MLIR_VERSION`. The same happens in `Dialects`, except that dialect modules are also generated.
- Methods in `API` and `Dialects` submodules (e.g. `arith`) dynamically dispatch to methods in versioned modules. The methods are programatically generated on precompile. An example of a generated method is shown here:
```julia
function mlirContextCreate(args...; kwargs...)
	version = MLIR_VERSION[]
	if v"14" <= version < v"15"
		v14.mlirContextCreate(args...; kwargs...)
	elseif v"15" <= version < v"16"
		v15.mlirContextCreate(args...; kwargs...)
	elseif v"16" <= version < v"17"
		v16.mlirContextCreate(args...; kwargs...)
	else
		throw(MLIRException("Unsupported MLIR version $version")
	end
end
```

# Open questions

- [x] ~Calls to versioned-functions should be performed through the `Dispatcher` (e.g. `IR.Context()` calls `API.Dispatcher.mlirContextCreate()`), or should we refactor the `API` module so that `MLIR.API` points to the `Dispatcher`?~
- [x] ~In `Dialects` module, we use `Dispatcher` to select the versioned dialect. The problem, since `Dispatcher` is not a `Module`, we no longer can import the dialect modules (unless you explicitly import a versioned module).~

### CODE REVIEW

File: src/Dialects/Dialects.jl [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    # generate version-less dialect modules
+    for (dialect, ops) in dialectops
+        mod = @eval module $dialect
+        using ...MLIR: MLIR_VERSION
+        using ..Dialects: v14, v15, v16
+        end
+
+        for op in ops
+            @eval mod function $op(args...; kwargs...)
+                if v"14" <= MLIR_VERSION[] < v"15"
```
Comment (jumerckx) [inline src/Dialects/Dialects.jl:39]: I'm wondering whether it's possible to throw a nice error if the version-less method is called with an `MLIR_VERSION` that doesn't support the operation? Currently, the behaviour is not unexpected but can be a bit confusing for someone who doesn't know how this is implemented.
```jl
julia> with(MLIR.MLIR_VERSION => v"14.0.0") do
           MLIR.Dialects.arith.addui_extended() # (exists in v16 but doesn't in v14)
       end
ERROR: UndefVarError: `addui_extended` not defined
Stacktrace:
  [1] getproperty
    @ ./Base.jl:31 [inlined]
  [2] addui_extended(::MLIR.IR.Value, ::Vararg{MLIR.IR.Value}; kwargs::@Kwargs{})
    @ MLIR.Dialects.arith ~/masterthesis/MLIR.jl/src/Dialects/Dialects.jl:40
  [3] addui_extended(::MLIR.IR.Value, ::Vararg{MLIR.IR.Value})
    @ MLIR.Dialects.arith ~/masterthesis/MLIR.jl/src/Dialects/Dialects.jl:38
  [4] (::var"#25#26")()
    @ Main ./REPL[27]:2
  [5] (::ScopedValues.var"#1#2"{var"#25#26"})()
    @ ScopedValues ~/.julia/packages/ScopedValues/aW0Zd/src/ScopedValues.jl:164
  [6] with_logstate(f::Function, logstate::Any)
    @ Base.CoreLogging ./logging.jl:515
  [7] with_logger
    @ ./logging.jl:627 [inlined]
  [8] enter_scope
    @ ~/.julia/packages/ScopedValues/aW0Zd/src/payloadlogger.jl:17 [inlined]
  [9] with(::Any, ::Pair{<:ScopedValue})
    @ ScopedValues ~/.julia/packages/ScopedValues/aW0Zd/src/ScopedValues.jl:163
 [10] top-level scope
    @ REPL[27]:1
```

**Review by jumerckx — APPROVED** [review]
I like this. The one thing I'm wondering is whether it's still useful to have the ability to set version and path statically, using preferences, as I imagine setting the scopedvalue instead doesn't really have a disadvantage?

I tested it out a bit and everything works great so this looks good to me. Curious to hear if there's objections by anyone else though, I don't really know what disadvantages this could bring? I've approved this but it might be wise to also get a set of more experienced eyes on this :)

File: src/Dialects/Dialects.jl (hunk 1 above)
Comment (mofeing) [reply src/Dialects/Dialects.jl:39]: I'm against putting a `try/catch` there to catch the `UndefVarError` but mmm... I think I can refactor it to have a mapping of function => version and generate the functions accordingly.

Comment (mofeing) [reply src/Dialects/Dialects.jl]: Now it should be shown like this (Also applied to functions in `API`):
```julia
julia> with(MLIR.MLIR_VERSION => v"14.0.0") do  
                  MLIR.Dialects.arith.addui_extended() # (exists in v16 but doesn't in v14)  
              end

ERROR: MLIR.Dialects.arith.addui_extended is not implemented for MLIR 14. You can find it in MLIR 16.
Stacktrace:
 [1] addui_extended(; kwargs::@Kwargs{})
   @ MLIR.Dialects.arith ~/Developer/MLIR.jl/src/Dialects/Dialects.jl:59
 [2] addui_extended()
   @ MLIR.Dialects.arith ~/Developer/MLIR.jl/src/Dialects/Dialects.jl:45
 [3] (::var"#7#8")()
   @ Main ./REPL[4]:2
 [4] (::ScopedValues.var"#1#2"{var"#7#8"})()
   @ ScopedValues ~/.julia/packages/ScopedValues/aW0Zd/src/ScopedValues.jl:164
 [5] with_logstate(f::Function, logstate::Any)
   @ Base.CoreLogging ./logging.jl:515
 [6] with_logger
   @ ./logging.jl:627 [inlined]
 [7] enter_scope
   @ ~/.julia/packages/ScopedValues/aW0Zd/src/payloadlogger.jl:17 [inlined]
 [8] with(::Any, ::Pair{<:ScopedValue})
   @ ScopedValues ~/.julia/packages/ScopedValues/aW0Zd/src/ScopedValues.jl:163
 [9] top-level scope
   @ REPL[4]:1
```

**Comment by mofeing** [comment]
> I like this. The one thing I'm wondering is whether it's still useful to have the ability to set version and path statically, using preferences, as I imagine setting the scopedvalue instead doesn't really have a disadvantage?

I added support for preferences in order for you to experiment with other libMLIR but yeah... probably doesn't make much sense now.

**Comment by jumerckx** [comment]
> I added support for preferences in order for you to experiment with other libMLIR

Oh, I was under the impression that it was possible to change the value of the scopedvalue but this doesn't seem possible without introducing a new scope using `with`.
In that case I think it's still valuable to have the preferences to allow this use-case, sorry for my confusion 😅

**Review by mofeing — COMMENTED** [review]
> Oh, I was under the impression that it was possible to change the value of the scopedvalue but this doesn't seem possible without introducing a new scope using `with`.
> In that case I think it's still valuable to have the preferences to allow this use-case, sorry for my confusion 😅 

Ah okay. It's not a problem. I just reverted the last commit.

**Pull request merged**